{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import get_data, analysis, utils, plotting, compute_comethylation, methyl_mut_burden\n",
    "import os\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dependency_f_dir = \"./dependency_files\"\n",
    "data_dir = \"./data\"\n",
    "mut_fn = os.path.join(data_dir, \"PANCAN_mut.tsv.gz\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reaad in mutations\n",
    "all_mut_df = get_data.get_mutations(mut_fn)\n",
    "all_mut_df['mutation'] = all_mut_df['chr'] + ':' + all_mut_df['start'].astype(str)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GoDMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in goDMC meQTLs\n",
    "meqtl_df = pd.read_parquet(\"/cellar/users/zkoch/methylation_and_mutation/data/meQTL/goDMC_meQTL/goDMC_meQTLs.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_246101/2764134249.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# split snp column on ':' and expand to get chr, start, and type\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mmeqtl_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'start'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'type'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeqtl_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'snp'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m':'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;31m# remove 'chr' from chr column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mmeqtl_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeqtl_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# create mutation column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/big_data/lib/python3.9/site-packages/pandas/core/accessor.py\u001b[0m in \u001b[0;36m__get__\u001b[0;34m(self, obj, cls)\u001b[0m\n\u001b[1;32m    179\u001b[0m             \u001b[0;31m# we're accessing the attribute of the class, i.e., Dataset.geo\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    180\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessor\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 181\u001b[0;31m         \u001b[0maccessor_obj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    182\u001b[0m         \u001b[0;31m# Replace the property with the accessor object. Inspired by:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    183\u001b[0m         \u001b[0;31m# https://www.pydanny.com/cached-property.html\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/big_data/lib/python3.9/site-packages/pandas/core/strings/accessor.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m    166\u001b[0m         \u001b[0;32mfrom\u001b[0m \u001b[0mpandas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstring_\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mStringDtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    167\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 168\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_inferred_dtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    169\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_categorical\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mis_categorical_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    170\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_string\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mStringDtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/big_data/lib/python3.9/site-packages/pandas/core/strings/accessor.py\u001b[0m in \u001b[0;36m_validate\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m    220\u001b[0m         \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"categories\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# categorical / normal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    221\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 222\u001b[0;31m         \u001b[0minferred_dtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfer_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mskipna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    223\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0minferred_dtype\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mallowed_types\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# split snp column on ':' and expand to get chr, start, and type\n",
    "meqtl_df[['chr', 'start', 'type']] = meqtl_df['snp'].str.split(':', expand=True)\n",
    "# remove 'chr' from chr column\n",
    "meqtl_df['chr'] = meqtl_df['chr'].str.replace('chr', '')\n",
    "# create mutation column\n",
    "meqtl_df['mutation'] = meqtl_df['chr'] + ':' + meqtl_df['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge on mutation to get all mutations that are meQTLs\n",
    "shared_mut_df = pd.merge(all_mut_df, meqtl_df, on='mutation', how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2837 of 2780312 somatic mutations are in the goDMC meqtls\n"
     ]
    }
   ],
   "source": [
    "print(\"{} of {} somatic mutations are in the goDMC meqtls\".format(len(shared_mut_df), len(all_mut_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0010203890786357791"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2837/2780312"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tcga meqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcga_meqtl_df = pd.read_csv(\"/cellar/users/zkoch/methylation_and_mutation/data/meQTL/pancan_tcga_meQTL/pancan_meQTL.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rm 'chr' from snp_position column\n",
    "tcga_meqtl_df['mutation'] = tcga_meqtl_df['snp_position'].str.replace('chr', '')\n",
    "# merge on mutation to get all mutations that are meQTLs\n",
    "tcga_merged = pd.merge(all_mut_df, tcga_meqtl_df, on='mutation', how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>reference</th>\n",
       "      <th>alt</th>\n",
       "      <th>DNA_VAF</th>\n",
       "      <th>mutation</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>snp</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>alleles</th>\n",
       "      <th>probes</th>\n",
       "      <th>probe_position</th>\n",
       "      <th>probe_gene</th>\n",
       "      <th>beta</th>\n",
       "      <th>status</th>\n",
       "      <th>r</th>\n",
       "      <th>p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-06-A5U1</td>\n",
       "      <td>1</td>\n",
       "      <td>152749036</td>\n",
       "      <td>152749036</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.24</td>\n",
       "      <td>1:152749036</td>\n",
       "      <td>LAML</td>\n",
       "      <td>rs149277953</td>\n",
       "      <td>chr1:152749036</td>\n",
       "      <td>GGGTGGT/G</td>\n",
       "      <td>cg24304714</td>\n",
       "      <td>chr1:152779374</td>\n",
       "      <td>LCE1C</td>\n",
       "      <td>-1.02</td>\n",
       "      <td>-4.57</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>1.430000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-13-2059</td>\n",
       "      <td>12</td>\n",
       "      <td>12482876</td>\n",
       "      <td>12482876</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.14</td>\n",
       "      <td>12:12482876</td>\n",
       "      <td>BLCA</td>\n",
       "      <td>rs113872597</td>\n",
       "      <td>chr12:12482876</td>\n",
       "      <td>CCTG/C</td>\n",
       "      <td>cg08615371</td>\n",
       "      <td>chr12:12503544</td>\n",
       "      <td>MANSC1</td>\n",
       "      <td>0.90</td>\n",
       "      <td>8.24</td>\n",
       "      <td>0.39</td>\n",
       "      <td>2.770000e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-13-2059</td>\n",
       "      <td>12</td>\n",
       "      <td>12482876</td>\n",
       "      <td>12482876</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.14</td>\n",
       "      <td>12:12482876</td>\n",
       "      <td>CESC</td>\n",
       "      <td>rs113872597</td>\n",
       "      <td>chr12:12482876</td>\n",
       "      <td>CCTG/C</td>\n",
       "      <td>cg08615371</td>\n",
       "      <td>chr12:12503544</td>\n",
       "      <td>MANSC1</td>\n",
       "      <td>0.87</td>\n",
       "      <td>6.19</td>\n",
       "      <td>0.36</td>\n",
       "      <td>2.320000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-13-2059</td>\n",
       "      <td>12</td>\n",
       "      <td>12482876</td>\n",
       "      <td>12482876</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.14</td>\n",
       "      <td>12:12482876</td>\n",
       "      <td>ESCA</td>\n",
       "      <td>rs113872597</td>\n",
       "      <td>chr12:12482876</td>\n",
       "      <td>CCTG/C</td>\n",
       "      <td>cg08615371</td>\n",
       "      <td>chr12:12503544</td>\n",
       "      <td>MANSC1</td>\n",
       "      <td>0.77</td>\n",
       "      <td>4.88</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2.740000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-13-2059</td>\n",
       "      <td>12</td>\n",
       "      <td>12482876</td>\n",
       "      <td>12482876</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.14</td>\n",
       "      <td>12:12482876</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>rs113872597</td>\n",
       "      <td>chr12:12482876</td>\n",
       "      <td>CCTG/C</td>\n",
       "      <td>cg08615371</td>\n",
       "      <td>chr12:12503544</td>\n",
       "      <td>MANSC1</td>\n",
       "      <td>0.95</td>\n",
       "      <td>10.67</td>\n",
       "      <td>0.44</td>\n",
       "      <td>5.650000e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>TCGA-VR-A8EY</td>\n",
       "      <td>16</td>\n",
       "      <td>87738888</td>\n",
       "      <td>87738888</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.50</td>\n",
       "      <td>16:87738888</td>\n",
       "      <td>TGCT</td>\n",
       "      <td>rs12447897</td>\n",
       "      <td>chr16:87738888</td>\n",
       "      <td>G/T</td>\n",
       "      <td>cg07873172</td>\n",
       "      <td>chr16:87712631</td>\n",
       "      <td>JPH3</td>\n",
       "      <td>0.41</td>\n",
       "      <td>4.55</td>\n",
       "      <td>0.38</td>\n",
       "      <td>1.280000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>TCGA-XE-AAOL</td>\n",
       "      <td>1</td>\n",
       "      <td>11982373</td>\n",
       "      <td>11982373</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.59</td>\n",
       "      <td>1:11982373</td>\n",
       "      <td>THCA</td>\n",
       "      <td>rs1137777</td>\n",
       "      <td>chr1:11982373</td>\n",
       "      <td>A/G</td>\n",
       "      <td>cg00299820</td>\n",
       "      <td>chr1:11943154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.37</td>\n",
       "      <td>7.35</td>\n",
       "      <td>0.32</td>\n",
       "      <td>8.980000e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>TCGA-XU-A936</td>\n",
       "      <td>12</td>\n",
       "      <td>125549956</td>\n",
       "      <td>125549956</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.64</td>\n",
       "      <td>12:125549956</td>\n",
       "      <td>LGG</td>\n",
       "      <td>rs41474647</td>\n",
       "      <td>chr12:125549956</td>\n",
       "      <td>C/T</td>\n",
       "      <td>cg03013422</td>\n",
       "      <td>chr12:125550350</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.52</td>\n",
       "      <td>7.67</td>\n",
       "      <td>0.34</td>\n",
       "      <td>1.030000e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>TCGA-XU-A936</td>\n",
       "      <td>12</td>\n",
       "      <td>125549956</td>\n",
       "      <td>125549956</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.64</td>\n",
       "      <td>12:125549956</td>\n",
       "      <td>LGG</td>\n",
       "      <td>rs41474647</td>\n",
       "      <td>chr12:125549956</td>\n",
       "      <td>C/T</td>\n",
       "      <td>cg21265540</td>\n",
       "      <td>chr12:125534110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.47</td>\n",
       "      <td>6.67</td>\n",
       "      <td>0.30</td>\n",
       "      <td>7.350000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>TCGA-XU-A936</td>\n",
       "      <td>12</td>\n",
       "      <td>125549956</td>\n",
       "      <td>125549956</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.64</td>\n",
       "      <td>12:125549956</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>rs41474647</td>\n",
       "      <td>chr12:125549956</td>\n",
       "      <td>C/T</td>\n",
       "      <td>cg16118089</td>\n",
       "      <td>chr12:125550542</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.47</td>\n",
       "      <td>4.41</td>\n",
       "      <td>0.34</td>\n",
       "      <td>1.980000e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>255 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample chr      start        end reference alt  DNA_VAF  \\\n",
       "0    TCGA-06-A5U1   1  152749036  152749036         G   T     0.24   \n",
       "1    TCGA-13-2059  12   12482876   12482876         C   T     0.14   \n",
       "2    TCGA-13-2059  12   12482876   12482876         C   T     0.14   \n",
       "3    TCGA-13-2059  12   12482876   12482876         C   T     0.14   \n",
       "4    TCGA-13-2059  12   12482876   12482876         C   T     0.14   \n",
       "..            ...  ..        ...        ...       ...  ..      ...   \n",
       "250  TCGA-VR-A8EY  16   87738888   87738888         G   T     0.50   \n",
       "251  TCGA-XE-AAOL   1   11982373   11982373         A   G     0.59   \n",
       "252  TCGA-XU-A936  12  125549956  125549956         C   T     0.64   \n",
       "253  TCGA-XU-A936  12  125549956  125549956         C   T     0.64   \n",
       "254  TCGA-XU-A936  12  125549956  125549956         C   T     0.64   \n",
       "\n",
       "         mutation cancer_type          snp     snp_position    alleles  \\\n",
       "0     1:152749036        LAML  rs149277953   chr1:152749036  GGGTGGT/G   \n",
       "1     12:12482876        BLCA  rs113872597   chr12:12482876     CCTG/C   \n",
       "2     12:12482876        CESC  rs113872597   chr12:12482876     CCTG/C   \n",
       "3     12:12482876        ESCA  rs113872597   chr12:12482876     CCTG/C   \n",
       "4     12:12482876        HNSC  rs113872597   chr12:12482876     CCTG/C   \n",
       "..            ...         ...          ...              ...        ...   \n",
       "250   16:87738888        TGCT   rs12447897   chr16:87738888        G/T   \n",
       "251    1:11982373        THCA    rs1137777    chr1:11982373        A/G   \n",
       "252  12:125549956         LGG   rs41474647  chr12:125549956        C/T   \n",
       "253  12:125549956         LGG   rs41474647  chr12:125549956        C/T   \n",
       "254  12:125549956        PAAD   rs41474647  chr12:125549956        C/T   \n",
       "\n",
       "         probes   probe_position probe_gene  beta  status     r       p-value  \n",
       "0    cg24304714   chr1:152779374      LCE1C -1.02   -4.57 -0.42  1.430000e-05  \n",
       "1    cg08615371   chr12:12503544     MANSC1  0.90    8.24  0.39  2.770000e-15  \n",
       "2    cg08615371   chr12:12503544     MANSC1  0.87    6.19  0.36  2.320000e-09  \n",
       "3    cg08615371   chr12:12503544     MANSC1  0.77    4.88  0.37  2.740000e-06  \n",
       "4    cg08615371   chr12:12503544     MANSC1  0.95   10.67  0.44  5.650000e-24  \n",
       "..          ...              ...        ...   ...     ...   ...           ...  \n",
       "250  cg07873172   chr16:87712631       JPH3  0.41    4.55  0.38  1.280000e-05  \n",
       "251  cg00299820    chr1:11943154        NaN  0.37    7.35  0.32  8.980000e-13  \n",
       "252  cg03013422  chr12:125550350       AACS  0.52    7.67  0.34  1.030000e-13  \n",
       "253  cg21265540  chr12:125534110        NaN  0.47    6.67  0.30  7.350000e-11  \n",
       "254  cg16118089  chr12:125550542       AACS  0.47    4.41  0.34  1.980000e-05  \n",
       "\n",
       "[255 rows x 19 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "big_data",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a1503193e5ffaa35c0c69857903e72fe32e1565e821b70ae44cce5264b7b1650"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
